Internal error on overflow in sum aggregate of decimal

If a sum aggregate of decimals overflows an internal error is thrown. This should instead be a user error and trigger `NUMERIC_VALUE_OUT_OF_RANGE` instead, similar to:
```
select sum(n) from (values 1, 9223372036854775807) s(n);
```

To reproduce:
```
select sum(n) from (values decimal '1', decimal '99999999999999999999999999999999999999') s(n);
```

The exception thrown is:
```
java.lang.ArithmeticException: Decimal overflow
	at io.trino.operator.aggregation.DecimalSumAggregation.outputDecimal(DecimalSumAggregation.java:123)
	at io.trino.$gen.sumAccumulator_20231227_105506_1994.evaluateFinal(Unknown Source)
	at io.trino.operator.aggregation.Aggregator.evaluate(Aggregator.java:93)
	at io.trino.operator.AggregationOperator.getOutput(AggregationOperator.java:161)
        ...
```

Thanks! I think I have it and will submit a new PR today.
Here's the output based on the updates suggested from the CLI:
```
trino> select sum(n) from (values decimal '1', decimal '99999999999999999999999999999999999999') s(n);
Query 20240102_145659_00000_8ug7d failed: Decimal overflow
```
...and the stack trace from TpchQueryRunner:
```
2024-01-02T09:57:03.628-0500	ERROR	stage-scheduler	io.trino.execution.StageStateMachine	Stage 20240102_145659_00000_8ug7d.0 failed
io.trino.spi.TrinoException: Decimal overflow
	at io.trino.operator.aggregation.DecimalSumAggregation.outputDecimal(DecimalSumAggregation.java:137)
	at io.trino.$gen.sumAccumulator_20240102_145703_114.evaluateFinal(Unknown Source)
	at io.trino.operator.aggregation.Aggregator.evaluate(Aggregator.java:93)
	at io.trino.operator.AggregationOperator.getOutput(AggregationOperator.java:161)
	at io.trino.operator.Driver.processInternal(Driver.java:398)
	at io.trino.operator.Driver.lambda$process$8(Driver.java:301)
	at io.trino.operator.Driver.tryWithLock(Driver.java:704)
	at io.trino.operator.Driver.process(Driver.java:293)
	at io.trino.operator.Driver.processForDuration(Driver.java:264)
	at io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:887)
	at io.trino.execution.executor.timesharing.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:187)
	at io.trino.execution.executor.timesharing.TimeSharingTaskExecutor$TaskRunner.run(TimeSharingTaskExecutor.java:565)
	at io.trino.$gen.Trino_testversion____20240102_145642_75.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.ArithmeticException: Decimal overflow
	at io.trino.spi.type.Decimals.throwIfOverflows(Decimals.java:301)
	at io.trino.operator.aggregation.DecimalSumAggregation.outputDecimal(DecimalSumAggregation.java:134)
	... 15 more
```

--- a/core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalSumAggregation.java
+++ b/core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalSumAggregation.java
@@ -14,6 +14,7 @@
 package io.trino.operator.aggregation;
 
 import io.trino.operator.aggregation.state.LongDecimalWithOverflowState;
+import io.trino.spi.TrinoException;
 import io.trino.spi.block.BlockBuilder;
 import io.trino.spi.block.Int128ArrayBlock;
 import io.trino.spi.block.Int128ArrayBlockBuilder;
@@ -30,6 +31,7 @@ import io.trino.spi.function.SqlType;
 import io.trino.spi.type.Decimals;
 import io.trino.spi.type.Int128;
 
+import static io.trino.spi.StandardErrorCode.NUMERIC_VALUE_OUT_OF_RANGE;
 import static io.trino.spi.type.Int128Math.addWithOverflow;
 
 @AggregationFunction("sum")
@@ -119,7 +121,7 @@ public final class DecimalSumAggregation
     {
         if (state.isNotNull()) {
             if (state.getOverflow() != 0) {
-                throw new ArithmeticException("Decimal overflow");
+                throw new TrinoException(NUMERIC_VALUE_OUT_OF_RANGE, "Decimal overflow");
             }
 
             long[] decimal = state.getDecimalArray();
@@ -127,8 +129,9 @@ public final class DecimalSumAggregation
 
             long rawHigh = decimal[offset];
             long rawLow = decimal[offset + 1];
-
-            Decimals.throwIfOverflows(rawHigh, rawLow);
+            if (Decimals.overflows(rawHigh, rawLow)) {
+                throw new TrinoException(NUMERIC_VALUE_OUT_OF_RANGE, "Decimal overflow");
+            }
             ((Int128ArrayBlockBuilder) out).writeInt128(rawHigh, rawLow);
         }
         else {

--- a/core/trino-main/src/main/java/io/trino/operator/scalar/DataSizeFunctions.java
+++ b/core/trino-main/src/main/java/io/trino/operator/scalar/DataSizeFunctions.java
@@ -63,7 +63,7 @@ public final class DataSizeFunctions
         try {
             return Decimals.valueOf(bytes);
         }
-        catch (ArithmeticException e) {
+        catch (TrinoException e) {
             throw new TrinoException(NUMERIC_VALUE_OUT_OF_RANGE, format("Value out of range: '%s' ('%sB')", dataSize, bytes));
         }
     }

--- a/core/trino-main/src/test/java/io/trino/operator/aggregation/TestDecimalSumAggregation.java
+++ b/core/trino-main/src/test/java/io/trino/operator/aggregation/TestDecimalSumAggregation.java
@@ -15,6 +15,7 @@ package io.trino.operator.aggregation;
 
 import io.trino.operator.aggregation.state.LongDecimalWithOverflowState;
 import io.trino.operator.aggregation.state.LongDecimalWithOverflowStateFactory;
+import io.trino.spi.TrinoException;
 import io.trino.spi.block.BlockBuilder;
 import io.trino.spi.block.Int128ArrayBlock;
 import io.trino.spi.block.VariableWidthBlockBuilder;
@@ -131,7 +132,7 @@ public class TestDecimalSumAggregation
 
         assertThat(state.getOverflow()).isEqualTo(1);
         assertThatThrownBy(() -> DecimalSumAggregation.outputDecimal(state, new VariableWidthBlockBuilder(null, 10, 100)))
-                .isInstanceOf(ArithmeticException.class)
+                .isInstanceOf(TrinoException.class)
                 .hasMessage("Decimal overflow");
     }
 

--- a/core/trino-spi/src/main/java/io/trino/spi/type/Decimals.java
+++ b/core/trino-spi/src/main/java/io/trino/spi/type/Decimals.java
@@ -305,10 +305,13 @@ public final class Decimals
         catch (Exception e) {
             throw new ArithmeticException("Decimal overflow");
         }
-        throwIfOverflows(result.getHigh(), result.getLow());
+        if (overflows(result.getHigh(), result.getLow())) {
+            throw new TrinoException(NUMERIC_VALUE_OUT_OF_RANGE, "Decimal overflow");
+        }
         return result;
     }
 
+    @Deprecated(forRemoval = true)
     public static void throwIfOverflows(long high, long low)
     {
         if (overflows(high, low)) {

